src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptions.java

```diff
@@ -596,6 +588,90 @@ public boolean apply(Object input) {
+        SingleNodeDisruption masterNodeDisruption = new LongGCDisruption(getRandom(), oldMasterNode);
+
+        // Save the majority side
+        final List<String> majoritySide = new ArrayList<>(nodes);
+        majoritySide.remove(oldMasterNode);
+
+        // Keep track of the masters that appear in the cluster state on both nodes on the majority side,
+        // only new_master may appear in here.
+        final Set<String> masters = new HashSet<>();
+        for (String node : majoritySide) {
+            internalCluster().getInstance(ClusterService.class, node).add(new ClusterStateListener() {
+                @Override
+                public void clusterChanged(ClusterChangedEvent event) {
+                    DiscoveryNode masterNode = event.state().getNodes().masterNode();
+                    if (masterNode != null) {
+                        masters.add(masterNode.name());
+                    }
+                }
+            });
+        }
+
+        internalCluster().setDisruptionScheme(masterNodeDisruption);
+        logger.info("freezing node [{}]", oldMasterNode);
+        masterNodeDisruption.startDisrupting();
+
+        // Wait for the majority side to get stable
+        ensureStableCluster(2, majoritySide.get(0));
+        ensureStableCluster(2, majoritySide.get(1));
+
+        // The old master node is frozen, but here we submit a cluster state update task that doesn't get executed,
+        // but will be queued and once the old master node un-freezes it gets executed.
+        // The old master node will send this update + the cluster state where he is flagged as master to the other
+        // nodes that follow the new master. These nodes should ignore this update.
+        internalCluster().getInstance(ClusterService.class, oldMasterNode).submitStateUpdateTask("sneaky-update", Priority.IMMEDIATE, new ClusterStateUpdateTask() {
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                return ClusterState.builder(currentState).build();
+            }
+
+            @Override
+            public void onFailure(String source, Throwable t) {
+                logger.warn("failure [{}]", t, source);
+            }
+        });
+
+        // Save the new elected master node
+        final String newMasterNode = internalCluster().getMasterName(majoritySide.get(0));
+        logger.info("new detected master node [{}]", newMasterNode);
+
+        // Stop disruption
+        logger.info("Unfreeze node [{}]", oldMasterNode);
+        masterNodeDisruption.stopDisrupting();
+
+        // Make sure that the end state is consistent on all nodes:
+        assertDiscoveryCompleted(nodes);
+        assertMaster(newMasterNode, nodes);
+
+        // During and after the disruption time only only one cluster state update may have occured that changed the master
+        // on nodes on the majority side.
```

Comment:
here we would validate that each node made two switches - one to remove the old master and one to accept the new.